File: Colorization/colorizeImage.py
#!/usr/bin/env python3
"""Colorize a still image with OpenCV DNN and an ONNX model."""

from __future__ import annotations

import argparse
import time
from pathlib import Path

import cv2 as cv

from colorization import DEFAULT_MODEL, colorize_frame, load_network, validate_output


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Colorize a grayscale or desaturated image with OpenCV DNN."
    )
    parser.add_argument(
        "--input",
        type=Path,
        default=Path(__file__).resolve().parent / "greyscaleImage.png",
        help="Input image (default: the bundled sample).",
    )
    parser.add_argument(
        "--model",
        type=Path,
        default=DEFAULT_MODEL,
        help="Path to colorization_eccv16.onnx.",
    )
    parser.add_argument(
        "--output",
        type=Path,
        default=Path("colorized-image.png"),
        help="Destination image.",
    )
    parser.add_argument(
        "--no-display",
        action="store_true",
        help="Do not open an OpenCV preview window.",
    )
    parser.add_argument(
        "--validate",
        action="store_true",
        help="Check the output dimensions, type, and predicted chroma.",
    )
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    image = cv.imread(str(args.input), cv.IMREAD_COLOR)
    if image is None:
        raise FileNotFoundError(f"Could not read input image: {args.input}")

    network = load_network(args.model)
    start = time.perf_counter()
    output, chroma_score = colorize_frame(image, network)
    elapsed = time.perf_counter() - start

    args.output.parent.mkdir(parents=True, exist_ok=True)
    if not cv.imwrite(str(args.output), output):
        raise RuntimeError(f"Could not write output image: {args.output}")
    if args.validate:
        validate_output(image, output, chroma_score)

    print(f"Saved {args.output}")
    print(f"Inference time: {elapsed:.3f} seconds")
    print(f"Mean predicted chroma: {chroma_score:.3f}")

    if not args.no_display:
        comparison = cv.hconcat([image, output])
        cv.imshow("Input | Colorized", comparison)
        cv.waitKey(0)
        cv.destroyAllWindows()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: Colorization/colorizeImage.cpp
// This code is written by Sunita Nayak at BigVision LLC.
// It is subject to the license terms in the LICENSE file in this folder.

#include "colorization.hpp"

#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>

#include <chrono>
#include <filesystem>
#include <iomanip>
#include <iostream>
#include <stdexcept>

int main(int argc, char** argv) {
    try {
        const colorization::CommonOptions options = colorization::parseOptions(
            argc,
            argv,
            "greyscaleImage.png",
            "colorized-image.png",
            false);

        const cv::Mat input = cv::imread(options.input.string(), cv::IMREAD_COLOR);
        if (input.empty()) {
            throw std::runtime_error(
                "Could not read input image: " + options.input.string());
        }

        cv::dnn::Net network = colorization::loadNetwork(options.model);
        const auto start = std::chrono::steady_clock::now();
        auto [output, chromaScore] =
            colorization::colorizeFrame(input, network);
        const std::chrono::duration<double> elapsed =
            std::chrono::steady_clock::now() - start;

        const std::filesystem::path parent = options.output.parent_path();
        if (!parent.empty()) {
            std::filesystem::create_directories(parent);
        }
        if (!cv::imwrite(options.output.string(), output)) {
            throw std::runtime_error(
                "Could not write output image: " + options.output.string());
        }
        if (options.validate) {
            colorization::validateOutput(input, output, chromaScore);
        }

        std::cout << "Saved " << options.output << '\n'
                  << "Inference time: " << std::fixed << std::setprecision(3)
                  << elapsed.count() << " seconds\n"
                  << "Mean predicted chroma: " << chromaScore << '\n';

        if (!options.noDisplay) {
            cv::Mat comparison;
            cv::hconcat(input, output, comparison);
            cv::imshow("Input | Colorized", comparison);
            cv::waitKey(0);
            cv::destroyAllWindows();
        }
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "Error: " << error.what() << '\n';
        return 1;
    }
}
